QR Scanner

ifisq opened the pull request:
## Description

Fixes #135 

- Adds QR scanner functionality to app.

## Todos

## Screenshots

I don't have a screenshot of this working since we don't have any usable QR codes yet. The only added page is a camera for the QR scanner.

ashayp22 commented on HIAPI/Models/Event.swift:
```diff
@@ -70,3 +70,18 @@ public struct EventCheckInStatus: Codable, APIReturnable {
+}
+
+public struct UserTracker: Codable {
+    public let userId: String
```
No longer a need for this since staff won't be scanning. Can be deleted ❌

ashayp22 commented on HIAPI/Services/EventService.swift:
```diff
@@ -28,6 +28,12 @@ public class EventService: BaseService {
         body["code"] = code
         return APIRequest<EventCheckInStatus>(service: self, endpoint: "checkin/", body: body, method: .POST)
     }
+    public static func track(eventId: String, userId: String) ->APIRequest<EventTrack> {
```
Same as above, can be deleted

ashayp22 commented on HackIllinois/FlowControllers/HITabBarController.swift:
```diff
@@ -41,10 +41,17 @@ class HITabBarController: UITabBarController {
-        codePopup.modalTransitionStyle = .crossDissolve
-        self.present(codePopup, animated: true, completion: nil)
+        if HIApplicationStateController.shared.isGuest {
+            let notEnabledAlert = UIAlertController(title: "This feature hasn't been enabled yet!", message: "", preferredStyle: .alert)
```
Maybe change the text to "You need to log out of your current account and log in as an attendee to check in to events."

ashayp22 commented on HackIllinois/ViewControllers/HIScanQRCodeViewController.swift:
```diff
@@ -0,0 +1,205 @@
+        containerView.addSubview(previewView)
+        view.addSubview(closeButton)
+        closeButton.addTarget(self, action: #selector(didSelectCloseButton(_:)), for: .touchUpInside)
+        closeButton.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 3).isActive = true
```
Could you make the closeButton a bit larger and farther from the edge so it is easer to tap?

ashayp22 reviewed (changes requested):
QR Scanner works perfectly, great work Aryan😄! 

I left some inline comments, but they should be quick to fix. 

After you finish up with closing the QR scanner after scanning a code, I can take a look at any new code and then you can merge ✅

ifisq commented:
@ashayp22 I've implemented your requested changes. I also updated the branch to the latest version of dev to resolve a merge conflict. Just wanted to let you know that for some reason, the latest version of dev isn't building on my MacBook properly, and I'm getting an error that it isn't able to load one of the assets you added. Here's a screenshot:
![image](https://user-images.githubusercontent.com/35982379/149643240-631548fa-c9b3-4e2d-8486-cb6572f7d658.png)

ashayp22 commented:
> @ashayp22 I've implemented your requested changes. I also updated the branch to the latest version of dev to resolve a merge conflict. Just wanted to let you know that for some reason, the latest version of dev isn't building on my MacBook properly, and I'm getting an error that it isn't able to load one of the assets you added. Here's a screenshot: ![image](https://user-images.githubusercontent.com/35982379/149643240-631548fa-c9b3-4e2d-8486-cb6572f7d658.png)

I deleted the Matching asset since we no longer will be using it. You can replace that line with:
`tabBarItem = UITabBarItem(title: "", image: #imageLiteral(resourceName: "leaderboard"), selectedImage: #imageLiteral(resourceName: "LeaderboardSelected"))`

ashayp22 reviewed (dismissed):
One more tiny change: can you make the color of the `closeButton` black? 

After that, feel free to go ahead and merge!

------------------------

Update: Sorry, one more thing: I noticed that if the user
1. Goes to the Profile page
2. Clicks the QR Scanner button and scans a QR code
3. Returns back to the Profile page after the QR scanner is closed.

The points aren't updated on the Profile Page. 

I think this is because `reloadProfile()` function in HIProfileViewController isn't called after the QR scanner is closed. Can you think of a way to call that function after the user successfully scans a QR code?

ifisq commented:
@ashayp22 i've added the requested changes, let me know if there's anything else i need to do before merging.

ashayp22 reviewed (approved):
Looks good, you can merge
